python/ray/data/_internal/execution/operators/output_splitter.py

```diff
@@ -0,0 +1,224 @@
+        return stats
+
+    def add_input(self, bundle, input_index) -> None:
+        if bundle.num_rows() is None:
+            raise ValueError("OutputSplitter requires bundles with known row count")
+        self._buffer.append(bundle)
+        self._dispatch_bundles()
+
+    def inputs_done(self) -> None:
+        if not self._equal:
+            # There shouldn't be any buffered data if we're not in equal split mode.
+            assert not self._buffer
+            return
+
+        # Otherwise:
+        # Need to finalize distribution of buffered data to output splits.
+        buffer_size = sum(b.num_rows() for b in self._buffer)
+        max_n = max(self._num_output)
+
+        # First calculate the min rows to add per output to equalize them.
+        allocation = [max_n - n for n in self._num_output]
+        remainder = buffer_size - sum(allocation)
+        # Invariant: buffer should always be large enough to equalize.
+        assert remainder >= 0, (remainder, buffer_size, allocation)
+
+        # Equally distribute remaining rows in buffer to outputs.
+        x = remainder // len(allocation)
+        allocation = [a + x for a in allocation]
+
+        # Execute the split.
+        for i, count in enumerate(allocation):
+            bundles = self._split_from_buffer(count)
+            for b in bundles:
+                b.output_split = i
+                self._output_queue.append(b)
+        self._buffer = []
+
+    def current_resource_usage(self) -> ExecutionResources:
+        return ExecutionResources(
+            object_store_memory=sum(b.size_bytes() for b in self._buffer)
+            + sum(b.size_bytes() for b in self._output_queue)
+        )
+
+    def progress_str(self) -> str:
+        if self._equal:
+            return f"{len(self._buffer)} buffered"
+        assert not self._buffer
+        return ""
+
+    def _dispatch_bundles(self) -> None:
+        # Dispatch all dispatchable bundles from the internal buffer.
+        # This may not dispatch all bundles when equal=True.
+        while self._buffer:
+            target_index = self._select_output_index()
+            target_bundle = self._pop_bundle_to_dispatch(target_index)
+            if self._can_safely_dispatch(target_index, target_bundle.num_rows()):
+                target_bundle.output_split = target_index
+                self._num_output[target_index] += target_bundle.num_rows()
+                self._output_queue.append(target_bundle)
+            else:
+                # Put it back and abort.
+                self._buffer.insert(0, target_bundle)
+                break
+
+    def _select_output_index(self) -> int:
+        # Greedily dispatch to the consumer with the least data so far.
+        i, _ = min(enumerate(self._num_output), key=lambda t: t[1])
+        return i
+
+    def _pop_bundle_to_dispatch(self, target_index: int) -> RefBundle:
+        # TODO implement locality aware bundle selection.
+        return self._buffer.pop(0)
```

Comment:
If we don't have any ordering constraints here, shouldn't we instead pop the smallest block from the buffer to make dispatching more likely? Otherwise you might have head-of-line blocking if the first block in the buffer is a large block.

Reply:
Once we do locality aware selection I think that will supercede this. In general I'm not too concerned about HOL blocking since for most reasonable workloads either (1) the number of blocks will be large or (2) we'll have a large buffer size. In these situations the bottleneck will be elsewhere.